File: Python/CRUD_GRANILARIDADE/crud.py
import json

#Menu principal

def menu():
    print("\n === MENU DE CONTATOS ===")
    print("1. Adicionar")
    print("2. Listar")
    print("3. Atualizar")
    print("4. Deletar")
    print("5. Sair")

#Função para escolher grupo

def escolher_grupo():
    #Pergunta ao usuário o grupo que deseja retornar
    print("\n Tipo de contato:")
    print("1. Aluno")
    print("2. Professor")

    opcao = input("Escolha: ")

    if opcao == "1":
        return "alunos"
    
    elif opcao == "2":
        return "professores"
    
    else:
        print("opção inválida!")
        return None
    
#função para ler o json
def ler_dados():
    with open("contatos.json", "r", encoding="utf-8") as arquivo:
        return json.load(arquivo)
    
#função salvar dados
def salvar_dados(dados):
    with open("contatos.json", "w", encoding="utf-8") as arquivo:
        json.dump(dados, arquivo, indent=4, ensure_ascii=False)

#Adicionar dados
def adicionar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    nome = input("Nome: ")
    telefone = input("Telefone: ")

    dados = ler_dados()

    dados[grupo].append({
        "nome": nome,
        "telefone": telefone
    })

    salvar_dados (dados)
    print("Contato adicionado com sucesso!")

#função listar contatos
def listar():
    grupo = escolher_grupo()

    if not grupo:
        return

    dados = ler_dados()
    print(f"\n Lista de {grupo.upper()}:")

    for index, contato in enumerate(dados[grupo], start=1):
        print(f"{index}. {contato['nome']} - {contato['telefone']}")

#atualizar
def atualizar():
    grupo = escolher_grupo()

    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("Index do contato: ")) -1

    if 0 <= index < len(dados[grupo]):
        nome = input("Novo nome: ")
        telefone = input("Novo telefone: ")

        dados[grupo][index] = {
            "nome": nome,
            "telefone": telefone
        }

        salvar_dados(dados)
        print("Contato atualizado com sucesso!")
    else:
        print("Indice inválido!")

#Função para excluir
def excluir():
    grupo = escolher_grupo()
    if not grupo:
        return
    
    dados = ler_dados()

    index = int(input("index do contato: ")) -1

    #verificar se o index é válido
    if 0 <= index < len(dados[grupo]):
        #remove o elemento da matriz
        dados[grupo].pop(index)

        salvar_dados(dados)
        print("Contato excluido com sucesso!")

    else:
        print("Opção inválida!")

#menu principal
def main():
    while True:
        menu()
        opcao = input("Escolha uma opção: ")
        if opcao == "1":
            adicionar()

        elif opcao == "2":
            listar()

        elif opcao == "3":
            atualizar()

        elif opcao == "4":
            excluir()

        elif opcao == "5":
            print("Saindo...")
            break

        else:
            print("Opção inválida!")

#executar o programa
main()

File: JS/Crud_granularidade/crud.js
const fs = require("fs");
const prompt = require("prompt-sync")();

function menu() {
    console.log("\nMenu de Contatos: ");
    console.log("1. Adicionar Contato");
    console.log("2. Listar Contatos");
    console.log("3. Atualizar Contato");
    console.log("4. Excluir Contato");
    console.log("5. Sair");
}

function escolherGrupo() {
    console.log("\n Tipos de Contatos: ");
    console.log("1. Aluno");
    console.log("2. Professor");

    const opcao = prompt("Escolha: ");

    if (opcao === "1") return "alunos";
    if (opcao === "2") return "professores";

    console.log("Opção inválida, retornando ao menu principal.");
    return null;
}

function lerDados() {
    const dados = fs.readFileSync("contatos.json", "utf-8");
    return JSON.parse(dados);
}

function salvarDados(dados) {
    fs.writeFileSync("contatos.json", JSON.stringify(dados, null, 2));
}

function adicionar() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const nome = prompt("Nome:");
    const telefone = prompt("Telefone:");

    const dados = lerDados();
    dados[grupo].push({ nome, telefone });

    salvarDados(dados);
    console.log("Contato adicionado com sucesso!");
}

function listar() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const dados = lerDados();

    console.log(`\n$Lista de ${grupo.toUpperCase()}:`);
    dados[grupo].forEach((contato, index) => {
        console.log(`${index + 1}. ${contato.nome} - ${contato.telefone}`);
    });
}

function atualizar() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const dados = lerDados();

    const index = parseInt(prompt("Index do contato:: ")) - 1;

    if (index >= 0 && index < dados[grupo].length){
        const nome = prompt("Novo nome:");
        const telefone = prompt("Novo telefone:");

        dados[grupo][index] = { nome, telefone };
        salvarDados(dados);

        console.log("Contato atualizado com sucesso!");
    } else {
        console.log("Índice inválido!");
    }
}

function excluir() {
    const grupo = escolherGrupo();
    if (!grupo) return;

    const dados = lerDados();
    const index = parseInt(prompt("Index do contato: ")) - 1;

    if (index >= 0 && index < dados[grupo].length) {
        dados[grupo].splice(index, 1);
        salvarDados(dados);

        console.log("Contato excluído com sucesso!");
    } else {
        console.log("Índice inválido!");
    }   
}

function main() {
    let opcao;

    do {
        menu();
        opcao = prompt("Escolha uma opção: ");
        switch (opcao) {
            case "1":
                adicionar();
                break;
            case "2":
                listar();
                break;
            case "3":
                atualizar();
                break;
            case "4":
                excluir();
                break;
            case "5":
                console.log("Saindo...");
                break;
            default:
                console.log("Opção inválida!");
        }
    } while (opcao !== "5");
}

main();

        

